### PR TITLE
[KYLIN-5387]: migrate cube planner1 to kylin5

### PR DESCRIPTION
**Pull request by liukun4515**
## Proposed changes

Dev design doc: https://docs.google.com/document/d/1wUNd8U1u-w8T-qQUReplPhDFkETj3AtFr_Shex501ls/edit?usp=sharing

- add config and checker to enable or disable the cube planner1 when build a new segment or refresh a segment. 
- add property of `@JsonProperty("layout_cost_based_pruned_list")` which is used to indicate whether the rulebaseindex has been recommended by the cost based algorithm.
- insert a  new stage of `COST_BASED_PLANNER` in `CUBING` job, and the new stage is used to create task to collect the statistics and run the cost based algorithm.
- use the `biginteger` instead of `long` to represent the cuboio
- migrate the algorithm from kylin3.1 to kylin5.0 without major changes, and just make some adaptive changes. For example, change the cubeio from `long` to `biginteger`
- migrate the IT/UT for the recommended algorithm from kylin3.1 to kylin5.0

## Branch to commit
- [ ] Branch **kylin3** for v2.x to v3.x
- [ ] Branch **kylin4** for v4.x
- [x] Branch **kylin5** for v5.x

## Types of changes

What types of changes does your code introduce to Kylin?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have created an issue on [Kylin's jira](https://issues.apache.org/jira/browse/KYLIN), and have described the bug/feature there in detail
- [ ] Commit messages in my PR start with the related jira ID, like "KYLIN-0000 Make Kylin project open-source"
- [ ] Compiling and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged

## Further comments

If this is a relatively large or complex change, kick off the discussion at user@kylin.apache.org or dev@kylin.apache.org by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: src/core-metadata/src/main/java/org/apache/kylin/metadata/cube/planner/algorithm/AbstractRecommendAlgorithm.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
+/*
```
Comment (liukun4515) [inline src/core-metadata/src/main/java/org/apache/kylin/metadata/cube/planner/algorithm/AbstractRecommendAlgorithm.java:1]: The files in this package are migrated from kylin3.1, the codes have not been changed.

File: src/core-metadata/src/main/java/org/apache/kylin/metadata/cube/planner/algorithm/genetic/BitsChromosome.java [hunk 2]
```diff
@@ -0,0 +1,117 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
```
Comment (liukun4515) [inline src/core-metadata/src/main/java/org/apache/kylin/metadata/cube/planner/algorithm/genetic/BitsChromosome.java:2]: The files in this package are migrated from kylin3.1, the codes have not been changed.

File: src/core-common/src/main/java/org/apache/kylin/common/KylinConfigBase.java [hunk 3]
```diff
@@ -3673,4 +3679,53 @@ public int getSecondStorageWaitLockTimeout() {
...
         return Boolean.parseBoolean(getOptional("kylin.source.ddl.enabled", FALSE));
     }
+
+    // ============================================================================
+    // Cost case cuboid Planner
+    // ============================================================================
+
+    public boolean enableCostBasedIndexPlanner() {
+        // If we enable the cost base cuboid planner, we will recommend subset of layouts from the rule.
+        return Boolean.parseBoolean(getOptional("kylin.index.costbased.enabled", FALSE));
```
Comment (liukun4515) [inline src/core-common/src/main/java/org/apache/kylin/common/KylinConfigBase.java:3689]: The default of cost based index is not enabled

File: src/kylin-it/src/test/java/org/apache/kylin/metadata/cube/planner/algorithm/ITAlgorithmTestBase.java [hunk 4]
```diff
@@ -0,0 +1,567 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
```
Comment (liukun4515) [inline src/kylin-it/src/test/java/org/apache/kylin/metadata/cube/planner/algorithm/ITAlgorithmTestBase.java:2]: The test cases are migrated from kylin3.1

**Comment by hit-lacus** [comment]
retest this please

**Comment by liukun4515** [comment]
@hit-lacus add the google doc to describe what I do in this PR

**Comment by liukun4515** [comment]
@hit-lacus please help to reset the ci

File: src/spark-project/spark-it/src/test/scala/org/apache/kylin/it/TestCubePlanner.scala [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+import org.apache.kylin.metadata.cube.model.NDataflowManager.NDataflowUpdater
+import org.apache.kylin.metadata.cube.model._
+import org.apache.kylin.metadata.realization.RealizationStatusEnum
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.SparderEnv
+import org.apache.spark.sql.common.{LocalMetadata, SparderBaseFunSuite}
+import org.apache.spark.sql.execution.adaptive.AdaptiveSparkPlanHelper
+import org.apache.spark.sql.execution.utils.SchemaProcessor
+
+class TestCubePlanner extends SparderBaseFunSuite
```
Comment (liukun4515) [inline src/spark-project/spark-it/src/test/scala/org/apache/kylin/it/TestCubePlanner.scala:33]: This is IT for cube planner:
1. basic test case
2. corner test case: delete one dimension after building one segment

File: src/spark-project/engine-spark/src/main/scala/org/apache/kylin/engine/spark/job/stage/build/CostBasedPlanner.scala [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+class CostBasedPlanner(jobContext: SegmentJob, dataSegment: NDataSegment, buildParam: BuildParam)
+  extends FlatTableAndDictBase(jobContext, dataSegment, buildParam) {
+  override def execute(): Unit = {
+    val (cost, sourceCount) = generateCostTable()
+    getRecommendedLayoutAndUpdateMetadata(cost, sourceCount)
+
+    val result = jobContext.updateIndexPlanIfNeed()
+    if (result) {
+      // update span tree and table desc with the new build job layouts
```
[the remark this answers is not in the document]
Comment (liukun4515) [reply src/spark-project/engine-spark/src/main/scala/org/apache/kylin/engine/spark/job/stage/build/CostBasedPlanner.scala:37]: partition and no-partition have different logic and different construct for the `AdaptiveSpanningTree` and `SegmentFlatTableDesc`, they are hard to be combined together.

On the original codebase, the code for partition and no-partition are separated, such as `PartitionMaterializedFactTableView` and `MaterializedFactTableView`.

[the remark this answers is not in the document]
Comment (liukun4515) [reply src/spark-project/engine-spark/src/main/scala/org/apache/kylin/engine/spark/job/stage/build/CostBasedPlanner.scala:37]: If you want to combine the logic for partition and no-partition, we can do this in the next follow-up PR.
But I find the logics in the `build`, `merge` and other `job` are all  divided into partition and non-partition

**Comment by hit-lacus** [comment]
Close this because it is merged in https://github.com/apache/kylin/pull/2089

**Pull request closed without merging**